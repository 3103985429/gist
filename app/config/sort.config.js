const SORT_ORDER_ASC = 'Ascending';
const SORT_ORDER_DESC = 'Descending';

module.exports = {
    SORT_ORDER_ASC,
    SORT_ORDER_DESC,
};
