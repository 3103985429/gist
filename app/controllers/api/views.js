const { handleError } = require('../../utils/requests');
const { getAllViews, getView, prepareView, scaffoldView } = require('../../utils/views');
const { getDataset } = require('../../utils/datasets');

module.exports.getAll = getAll;
module.exports.get = get;
module.exports.getDetailed = getDetailed;

async function getAll(req, res) {
    try {
        const views = await getAllViews(req.user);

        res.json(views);
    } catch (err) {
        handleError(err, res);
    }

}

async function get(req, res) {
    try {
        const view = await getView(req.params.id);
        const readyView = prepareView(view);

        res.json(readyView);
    } catch (err) {
        handleError(err, res);
    }
}

async function getDetailed(req, res) {
    try {
        const view = await getView(req.params.id);
        const dataset = await getDataset(req.params.datasetId);
        const preparedView = prepareView(view);
        const scaffoldedView = scaffoldView(preparedView, dataset);

        res.json(scaffoldedView);
    } catch (err) {
        handleError(err, res);
    }
}
