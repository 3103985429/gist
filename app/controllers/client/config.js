// legacy
const { cond, constant, eq, kebabCase, stubTrue } = require('lodash');

const {
    VIEW_TYPE_BAR_CHART,
    VIEW_TYPE_PIE_SET,
    VIEW_TYPE_SCATTERPLOT,
    VIEW_TYPE_TIMELINE,
    VIEW_TYPE_TREEMAP,
} = require('../../config/views.config');

const PAGE_SIZE_BAR_CHART = 1000;
const PAGE_SIZE_PIE_SET = 50;
const PAGE_SIZE_SCATTERPLOT = 1000;
const PAGE_SIZE_TIMELINE = 20;
const PAGE_SIZE_TREEMAP = 1000;
const PAGE_SIZE_DEFAULT = 200;

function defaultPageSize(type) {
    const equals = a => b => eq(a, b);

    return cond([
        [equals(VIEW_TYPE_BAR_CHART), constant(PAGE_SIZE_BAR_CHART)],
        [equals(VIEW_TYPE_PIE_SET), constant(PAGE_SIZE_PIE_SET)],
        [equals(VIEW_TYPE_SCATTERPLOT), constant(PAGE_SIZE_SCATTERPLOT)],
        [equals(VIEW_TYPE_TIMELINE), constant(PAGE_SIZE_TIMELINE)],
        [equals(VIEW_TYPE_TREEMAP), constant(PAGE_SIZE_TREEMAP)],
        [stubTrue, constant(PAGE_SIZE_DEFAULT)],
    ])(type);
}

const numberOperations = ['ToInteger', 'ToFloat', 'ToPercent', 'ToCurrency'];

const groupSizes = {
    timeline: [19, 11],
    pieSet: 50,
};
const defaultDateFormat = 'MM/DD/YYYY';

const AGGREGATE_BY_DEFAULT_COLUMN_NAME = 'Number of Items';

const AGGREGATE_BY_DISABLED_COLUMN_NAME = 'None';

const GROUP_BY_DATE_DISABLED = 'None';

const isDate = (dataSourceDescription, columnName) => {
    if (dataSourceDescription.raw_rowObjects_coercionScheme) {
        const coercion = dataSourceDescription.raw_rowObjects_coercionScheme;
        if (coercion[columnName] && coercion[columnName].operation === 'ToDate') {
            return true;
        } else {
            // if there are any custom fields to process, recurse through them
            for (let i = 0; i < dataSourceDescription.customFieldsToProcess.length; i++) {
                const mergedFields = dataSourceDescription.customFieldsToProcess[i].fieldsToMergeIntoArray;
                const fieldName = dataSourceDescription.customFieldsToProcess[i].fieldName;
                if (fieldName === columnName) {
                    return isDate(dataSourceDescription, mergedFields[i]);
                }
            }
        }
    }
    return false;
};

const isNumber = (dataSourceDescription, columnName) => {
    if (dataSourceDescription.raw_rowObjects_coercionScheme) {
        const coercion = dataSourceDescription.raw_rowObjects_coercionScheme;
        if (coercion[columnName] && numberOperations.includes(coercion[columnName].operation)) {
            return true;
        }
    }
    return false;
};

const formatDefaultView = view => kebabCase(view);

const hasMoreThanOneView = views => {
    let numberOfViews = 0;
    for (let view in views) {
        if (views[view].visible === true) {
            numberOfViews++;
        }
        if (numberOfViews > 1) {
            return true;
        }
    }
    return false;
};

// todo: merge with calculatePageRanges?
const calculateGalleryPageRanges = numResults => {
    let pageRanges = [];

    if (numResults > 200) {
        pageRanges = [5, 10, 20, 50, 100, 200];
    } else if (numResults > 100) {
        pageRanges = [5, 10, 20, 50, 100];
    } else if (numResults > 50) {
        pageRanges = [5, 10, 20, 50];
    } else if (numResults > 20) {
        pageRanges = [5, 10, 20];
    } else if (numResults > 10) {
        pageRanges = [5, 10];
    } else if (numResults > 5) {
        pageRanges = [5];
    }

    if (!pageRanges.includes(numResults) && numResults < 200) {
        pageRanges.push(numResults);
    }

    return pageRanges;
};

const userRoles = {
    visualizationEditor: 'visualizationEditor',
    admin: 'admin',
    editor: 'editor',
    viewer: 'viewer',
};

const PERCENT_UNIT = '%';

module.exports = {
    defaultPageSize,
    groupSizes,
    defaultDateFormat,
    AGGREGATE_BY_DEFAULT_COLUMN_NAME,
    AGGREGATE_BY_DISABLED_COLUMN_NAME,
    GROUP_BY_DATE_DISABLED,
    PERCENT_UNIT,
    isDate,
    isNumber,
    formatDefaultView,
    hasMoreThanOneView,
    calculateGalleryPageRanges,
    userRoles,
};
