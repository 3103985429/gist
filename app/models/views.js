var winston = require('winston');
const _ = require('lodash');

var mongoose_client = require('./mongoose_client');
var mongoose = mongoose_client.mongoose;
var Schema = mongoose.Schema;
var View_scheme = Schema({
    name: String,
    settings: { type: Array, default: [] },

    // be sure to call .markModified(path) on the model before saving if you update this Mixed property via Mongoose
    lookup: Schema.Types.Mixed,

    displayAs: String,
    icon: String,
    thumbnail: String,
    scaffold: { type: Object, default: {} },
    // _team: {type: Schema.Types.ObjectId, ref: 'Team'}
});

var view = mongoose.model('View', View_scheme);
var Team = require('./teams');

view.getAllBuiltInViews = function (fn) {
    view.find({})
        .select({ _id: 0, name: 1 })
        .exec(function (err, builtInViews) {
            if (err) {
                fn(err);
            } else {
                if (!builtInViews.length) {
                    winston.warn('List of built-in views is empty.');
                }
                builtInViews.forEach((view) => {
                    view.name = _.kebabCase(view.name);
                });
                fn(null, builtInViews);
            }
        });
};

view.getAllBuiltInViewsAsObject = async function () {
    const views = await view
        .find({})
        .select({ _id: 0, name: 1, displayAs: 1 });

    // Create object of view display names with names as keys
    return _.transform(views, (result, { name, displayAs }) => {
        result[name] = displayAs;
    }, {});
};

view.getAllCustomViews = function (fn) {
    var customViews = [];
    Team.find({ isEnterprise: true }, { subdomain: 1 })
        .exec(function (err, allEnterprise) {
            if (allEnterprise) {
                allEnterprise.forEach((team) => {
                    customViews.push(team.subdomain);
                });
            }
            fn(err, customViews);

        });

    // var customViews = ["atlas","insight","rhodiumgroup"];
    // fn(null,customViews);
};

module.exports = view;
