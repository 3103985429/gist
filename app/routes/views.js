const winston = require('winston');
const express = require('express');
const ConnectSequence = require('connect-sequence');
const { isNil } = require('lodash');
const router = express.Router();

const slowDown = require('../utils/slowDown');

// Express middleware
const { validate } = require('../libs/middleware/views/validate');
const { ensureAuthorized } = require('../libs/middleware/ensure-authorized');
const { puppeteerJSAuth } = require('../libs/middleware/puppeteer-js-auth');
const { loadPreview } = require('../libs/middleware/views/load-preview');
const { addSourceKey } = require('../libs/middleware/views/add-source-key');
const { addCdn } = require('../libs/middleware/views/add-cdn');
const { addRevision } = require('../libs/middleware/views/add-revision');
const { addMeta } = require('../libs/middleware/views/add-meta');
const { exists } = require('../libs/middleware/views/exists');
const { addViewType } = require('../libs/middleware/views/add-view-type');
const { addUser } = require('../libs/middleware/views/add-user');
const { addSharedPage } = require('../libs/middleware/views/add-shared-page');
const { addSourceAndSample } = require('../libs/middleware/views/add-source-and-sample');
const { addStories } = require('../libs/middleware/views/add-stories');
const { addGroupBy } = require('../libs/middleware/views/add-group-by');
const { addChartBy } = require('../libs/middleware/views/add-chart-by');
const { addSegmentByBy } = require('../libs/middleware/views/add-segment-by');
const { addAggregateBy } = require('../libs/middleware/views/add-aggregate-by');
const { addAxes } = require('../libs/middleware/views/add-axes');
const { addRadius } = require('../libs/middleware/views/add-radius');
const { addSortBy } = require('../libs/middleware/views/add-sort-by');
const { addStackBy } = require('../libs/middleware/views/add-stack-by');
const { addMapBy } = require('../libs/middleware/views/add-map-by');
const { addGroupSize } = require('../libs/middleware/views/add-group-size');
const { addSecondaryCol } = require('../libs/middleware/views/add-secondary-col');
const { addFilter } = require('../libs/middleware/views/add-filter');
const { addColFilter } = require('../libs/middleware/views/add-col-filter');
const { addSearch } = require('../libs/middleware/views/add-search');
const { addPagination } = require('../libs/middleware/views/add-pagination');
const { addImageMeta } = require('../libs/middleware/views/add-image-meta');
const { addRoutePathBase } = require('../libs/middleware/views/add-route-path-base');
const { addBarsAndType } = require('../libs/middleware/views/add-bars-and-types');
const { addTimeValue } = require('../libs/middleware/views/add-time-value');
const { render: renderView } = require('../libs/middleware/views/render');
const { render: renderObjectDetails } = require('../libs/middleware/object-details/render');
const { addAreas } = require('../libs/middleware/views/add-areas');
const { addLines } = require('../libs/middleware/views/add-lines');
const { addPies } = require('../libs/middleware/views/add-pies');
const { addClickThroughView } = require('../libs/middleware/views/add-click-through-view');
const { addInstanceId } = require('../libs/middleware/views/add-instance-id');
const { addObjectDetailsPagination } = require('../libs/middleware/views/add-object-details-pagination');
const { addDataPrepOptions } = require('../libs/middleware/views/add-data-prep-options');

// Models
const View = require('../models/views');
const { VIEW_TYPE_GROUPED_GALLERY } = require('../config/views.config');

// Set up route handling for all built in views
// Refactored routes: initial page render - does NOT include graph data //
const middlewareHead = [
    addSharedPage,
    validate,
    addSourceKey,
    addCdn,
    addRevision,
    addUser,
    puppeteerJSAuth,
    ensureAuthorized,
];
const middlewareTail = [
    loadPreview,
    exists,
    addFilter,
    addColFilter,
    addSourceAndSample,
    addStories,
    addSortBy,
    addGroupBy,
    addChartBy,
    addSegmentByBy,
    addAggregateBy,
    addAxes,
    addRadius,
    addStackBy,
    addMapBy,
    addGroupSize,
    addSecondaryCol,
    addSearch,
    addPagination,
    addImageMeta,
    addRoutePathBase,
    addBarsAndType,
    addTimeValue,
    addAreas,
    addLines,
    addPies,
    addClickThroughView,
    addInstanceId,
    renderView,
];

const objectDetailMiddlewareHead = [
    addSharedPage,
    validate,
    addSourceKey,
    addRevision,
    addUser,
    ensureAuthorized,
    loadPreview,
];

const objectDetailMiddlewarePaginated = [
    addViewType,
    addMeta(),
    exists,
    addFilter,
    addColFilter,
    addSourceAndSample,
    addStories,
    addSortBy,
    addGroupBy,
    addChartBy,
    addAggregateBy,
    addAxes,
    addRadius,
    addStackBy,
    addMapBy,
    addGroupSize,
    addSecondaryCol,
    addSearch,
    addDataPrepOptions,
    addObjectDetailsPagination,
];

const objectDetailMiddlewareTail = [
    addRoutePathBase,
    renderObjectDetails,
];

function addPaginatedMiddlewares(req, res, next) {
    const isPaginatedSet = ({ query: { objectIndex } }) => !isNil(objectIndex);

    new ConnectSequence(req, res, next).appendIf(isPaginatedSet, ...objectDetailMiddlewarePaginated).run();
}

View.getAllBuiltInViews(function (err, defaultViews) {
    if (err) {
        winston.error('Error getting default views to bind for routes: ', err);
    } else {
        const slowAfterReloads = 15;
        const windowTime = 60 * 1000;

        defaultViews.push({ name: VIEW_TYPE_GROUPED_GALLERY });
        // views
        defaultViews.forEach(function (view) {
            router.get('/:source_key/' + view.name, slowDown(slowAfterReloads, windowTime), middlewareHead, addMeta(view), middlewareTail);
        });
        // object detail
        router.get('/:source_key/:object_id', slowDown(slowAfterReloads, windowTime), objectDetailMiddlewareHead, addPaginatedMiddlewares, objectDetailMiddlewareTail);
        // viz with no view specified
        router.get('/:source_key', slowDown(slowAfterReloads, windowTime), middlewareHead, addMeta());
    }
});

module.exports = {
    router,
    objectDetailMiddlewareHead,
    addPaginatedMiddlewares,
    objectDetailMiddlewareTail,
};
