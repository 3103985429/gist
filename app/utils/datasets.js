const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
