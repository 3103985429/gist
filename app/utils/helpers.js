const { isEqual, trim, toString } = require('lodash');

module.exports.equals = a => b => isEqual(a, b);

module.exports.isEmbed = (query) => query.embed === 'true' || query.embed === true;

module.exports.formatColumnName = (value) => trim(toString(value).replace(/\./g, '_'));

module.exports.formatRawValues = (value) => trim(toString(value).replace(/’/g, '\'').replace(/”/g, '"'));

module.exports.addSlugIndex = (slugs, item, tilteProp = 'title') => {
    if (!slugs.length) {
        return;
    }

    let i = 2;
    const slugList = slugs.map((item) => item.slug);
    const max = i + slugList.length + 1;

    for (i; i < max; i++) {
        const newSlug = `${item.slug}-${i}`;

        if (!slugList.includes(newSlug)) {
            item.slug = newSlug;
            item[tilteProp] = `${item[tilteProp]} ${i}`;

            break;
        }
    }
};
