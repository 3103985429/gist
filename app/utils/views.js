const { isUndefined, map } = require('lodash');

const Views = require('../models/views');
const { getUser } = require('./users');
const { setDefaultsIfNoneExist } = require('../controllers/api/dataset/mandatory-settings-helpers');
const { mandatorySettings } = require('../../config/mandatory-settings');
const { UserError } = require('../libs/system/errors');

module.exports = {
    getAllViews,
    getAllViewsSettings,
    getView,
    prepareView,
    scaffoldView,
};

/**
 * @param {String|ObjectId} userId
 * @returns {Promise<Object>}
 */
async function getAllViews(userId) {
    const user = await getUser(userId);

    return await Views
        .find({
            $or: [
                { _team: { $exists: false } },
                { _team: user.defaultLoginTeam },
            ],
        })
        .select('_id name displayAs icon _team thumbnail scaffold')
        .lean()
        .exec();
}

/**
 * @param {String|ObjectId} userId
 * @returns {Promise<Object>}
 */
async function getAllViewsSettings(userId) {
    const user = await getUser(userId);

    return await Views
        .find({
            $or: [
                { _team: { $exists: false } },
                { _team: user.defaultLoginTeam },
            ],
        })
        .select('_id name displayAs settings')
        .lean()
        .exec();
}

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getView(id) {
    return await Views
        .findById(id)
        .lean()
        .exec();
}

/**
 * @param {Object} view
 * @returns {Object}
 */
function prepareView(view) {
    const { settings = [], scaffold = {} } = view;

    settings.forEach(({ inputType, selectExcludeBy }) => {
        if (['menu', 'checkbox'].includes(inputType) && selectExcludeBy && isUndefined(scaffold[selectExcludeBy])) {
            scaffold[selectExcludeBy] = [];
        }
    });

    return Object.assign({}, view, { settings, scaffold });
}

/**
 * @param {Object} view
 * @param {Object<Object>} raw_rowObjects_coercionScheme
 * @param {Array<String>} fe_fieldDisplayOrder
 * @param {Object<Boolean>} fe_excludeFields
 * @returns {Object}
 * @throws {Error}
 */
function scaffoldView(view, { raw_rowObjects_coercionScheme, fe_fieldDisplayOrder, fe_excludeFields }) {
    const columns = map(raw_rowObjects_coercionScheme, ({ sample }, name) => ({ name, sample }));
    const { message, ...scaffold } = setDefaultsIfNoneExist(
        columns,
        raw_rowObjects_coercionScheme,
        fe_fieldDisplayOrder,
        mandatorySettings[view.name],
        fe_excludeFields,
        { ...view.scaffold, visible: true },
        view.name,
    );

    if (message) {
        throw new UserError(message);
    }

    return { ...view, scaffold };
}
