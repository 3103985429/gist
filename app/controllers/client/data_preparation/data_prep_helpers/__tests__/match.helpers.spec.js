import { pokemon1 } from '../../../../../../tests/config/datasets';
import { VIEW_TYPE_GALLERY, VIEW_TYPE_REGIONAL_MAP } from '../../../../../config/views.config';
import { getMatch } from '../match.helpers';


describe('project helpers', () => {
    let dataSource;

    beforeEach(() => {
        dataSource = {};
        Object.assign(dataSource, pokemon1, {
            fe_views: {
                views: {
                    gallery: {
                        defaultSortByColumnName: 'Image URL',
                        visible: true,
                    },
                    regionalMap: {
                        regionField: 'Name',
                        region: 'EU',
                    },
                },
                default_view: 'gallery',
            },
        });
    });

    describe('getMatch', () => {
        it('should return default value', () => {
            const result = getMatch(VIEW_TYPE_GALLERY, dataSource);
            expect(result).toMatchSnapshot();
        });

        it('should return match for regional map', () => {
            const result = getMatch(VIEW_TYPE_REGIONAL_MAP, dataSource);
            expect(result).toMatchSnapshot();
        });
    });
});
