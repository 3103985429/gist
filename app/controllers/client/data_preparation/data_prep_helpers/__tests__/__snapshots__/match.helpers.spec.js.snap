// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project helpers getMatch should return default value 1`] = `
Object {
  "rowParams": Object {
    "$exists": true,
  },
}
`;

exports[`project helpers getMatch should return match for regional map 1`] = `
Object {
  "rowParams.Name": Object {
    "$in": Array [
      "Albania",
      "Andorra",
      "Armenia",
      "Austria",
      "Azerbaijan",
      "Belgium",
      "Bulgaria",
      "Bosnia and Herzegovina",
      "Belarus",
      "Switzerland",
      "Cyprus",
      "Czechia",
      "Germany",
      "Denmark",
      "Spain",
      "Estonia",
      "Finland",
      "France",
      "United Kingdom",
      "Georgia",
      "Greece",
      "Croatia",
      "Hungary",
      "Ireland",
      "Iceland",
      "Italy",
      "Kosovo",
      "Liechtenstein",
      "Lithuania",
      "Luxembourg",
      "Latvia",
      "Monaco",
      "Moldova",
      "Macedonia",
      "Montenegro",
      "Netherlands",
      "Norway",
      "Poland",
      "Portugal",
      "Romania",
      "Russia",
      "San Marino",
      "Serbia",
      "Slovakia",
      "Slovenia",
      "Sweden",
      "Turkey",
      "Ukraine",
      "Vatican",
    ],
  },
}
`;
